### DISCUSSION TITLE
saveimaged saving 2D masks in nifti format flipped

### DISCUSSION THREAD

**Discussion by OmarAshkar** [Project-MONAI/MONAI#6490] [Q&A] [answered]
**Describe the bug**
I am trying to save masks from hovernet pipeline similar to the tutorial here: https://github.com/Project-MONAI/tutorials/blob/main/pathology/hovernet/inference.py. When reloading the same masks by nibabel.load(), I get the images flipped (y,x) instead of (x,y), then flipped up.

I have to have a 32bit format to capture instance segmentation mask, so png didn't work for me and tif produces large files, so Nifti is what I need now.

**To Reproduce**
1. run the hovernet pipeline from the tutorial for link above
2. load one of the outputs by nibabel.load(), then .get_fdata()

**Expected behavior**
It should have been in the same dimensions of original image.

**Additional context**
This is crucial in this context, as png format supports only uint8, and tif files are huge and cannot be compressed from saveimaged.

Thanks!

**Comment by wyli** [comment]
could you please provide an example of the saving/loading?

**Reply by OmarAshkar** [reply]
Hi @wyli 

My current pre-transform looks like:

```
    pre_transforms = Compose(
        [
            LoadImaged(keys="image", reader=PILReader, converter=lambda x: x.convert("RGB")),
            EnsureChannelFirstd(keys="image"), 
            CastToTyped(keys="image", dtype=torch.float32),
            ScaleIntensityRanged(keys="image", a_min=0.0, a_max=255.0, b_min=0.0, b_max=1.0, clip=True),
        ]
    )
```

my post-transform process ends with 
```
            SaveImaged(
                keys="instance_map",
                meta_keys="image_meta_dict",
                output_ext=".nii.gz",
                output_dir=output_dir,
                output_postfix="instance_map",
                output_dtype="uint32",
                separate_folder=False
            ),
```

The save mapped is flipped wrt the original image. It is probably a hovernet processing problem.

**Reply by wyli** [reply]
thanks, this is 'flipped' mainly to ensure a consistent load/save round trip, this example is with monai 1.2.0rc6:

```py
import os

import torch

import monai
import monai.transforms as mt
from monai.data import PILReader

filename = os.path.join(".", "MONAI-logo_color.png")
monai.apps.download_url("https://monai.io/assets/img/MONAI-logo_color.png", filepath=filename)

data = {"image": filename}

pre_transforms = mt.Compose(
    [
        mt.LoadImaged(keys="image", reader=PILReader, converter=lambda x: x.convert("RGB"), reverse_indexing=False),
        mt.EnsureChannelFirstd(keys="image"),
        mt.CastToTyped(keys="image", dtype=torch.float32),
        mt.ScaleIntensityRanged(keys="image", a_min=0.0, a_max=255.0, b_min=0.0, b_max=1.0, clip=True),
    ]
)

saver = mt.SaveImaged(
    keys="image",
    resample=False,
    output_ext=".nii.gz",
    output_dir="./",
    output_postfix="img",
    output_dtype=float,
    separate_folder=False,
    savepath_in_metadict=True,
)

out = pre_transforms(data)
print(out["image"].shape)
saver(out)
saved = mt.LoadImage(image_only=True, ensure_channel_first=True)(out["image"].meta["saved_to"])
print(saved.shape)


```

output:
```
torch.Size([3, 100, 31])
2023-05-09 16:20:09,042 INFO image_writer.py:197 - writing: MONAI-logo_color_img.nii.gz
torch.Size([3, 100, 31, 1])
```

**Reply by OmarAshkar** [reply]
Hi @wyli  Thank you for your response. But I still didn't get it. I have images that I feed to this script and the outputs are flipped on saving nifti. This is not the case if I use PIL save as png. If you can elaborate please would be appreciated. Thanks!

**Reply by KumoLiu** [reply]
Hi @OmarAshkar, I just tried the Hovernet inference script and didn't find x,y has been flipped. Could you please share more information on how to find x,y has been flipped?
```
import matplotlib.pyplot as plt
import nibabel as nib

pre_transforms = mt.Compose(
    [
        mt.LoadImaged(keys="image", reader=PILReader, converter=lambda x: x.convert("RGB")),
        mt.EnsureChannelFirstd(keys="image"),
        mt.CenterSpatialCropd(keys="image", roi_size=(900, 800)),
        mt.CastToTyped(keys="image", dtype=torch.float32),
        mt.ScaleIntensityRanged(keys="image", a_min=0.0, a_max=255.0, b_min=0.0, b_max=1.0, clip=True),
    ]
)

out = pre_transforms({"image":  image_path})
data = nib.load(inference_instance_map_path)

print(out["image"].shape)  # --> torch.Size([3, 900, 800])
print(data.get_fdata().shape) # --> (900, 800)
```
Thanks!

**Reply by OmarAshkar** [reply]
Hi @KumoLiu Thank you for putting the time to look into that. To reproduce the error using the HoverNet tutorial, simply open one of the saved nii masks and compare it visually to the corresponding images. 

Also, I think image shape is the same. It is just rotated in a certain way. 

Thanks!

**Reply by KumoLiu** [reply, marked as answer]
Hi @OmarAshkar, yes I just open one of the saved predictions and compare it visually to the corresponding images. They are not rotated.
```
print(out["image"].shape)  # --> torch.Size([3, 900, 800])
print(data.get_fdata().shape) # --> (900, 800)
plt.subplot(1,2,1)
plt.imshow(out["image"].permute(1,2,0))
plt.subplot(1,2,2)
plt.imshow(data.get_fdata())
```
![image](https://github.com/Project-MONAI/MONAI/assets/55491388/b835ce87-daee-480f-8514-a61af6505445)

Thanks!

**Reply by OmarAshkar** [reply]
@KumoLiu I don't know what am I missing here. I will try to check my transformers to see which one causes that. But I can repreduce it from my end by switching ".nii" to ".png", things work as expected.
